### PR TITLE
Implementation of Tag.

### PR DESCRIPTION
**Pull request by wenchaozhang-123**
fix #ISSUE_Number

---

### Change logs

_Describe your change clearly, including what problem is being solved or what feature is being added._

_If it has some breaking backward or forward compatibility, please clary._

### Why are the changes needed?

_Describe why the changes are necessary._

### Does this PR introduce any user-facing change?

_If yes, please clarify the previous behavior and the change this PR proposes._

### How was this patch tested?

_Please detail how the changes were tested, including manual tests and any relevant unit or integration tests._

### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [ ] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [ ] Sign the Contributor License Agreement as prompted for your first-time contribution(*One-time setup*).
- [ ] Learn the [coding contribution guide](https://cloudberrydb.org/contribute/code), including our code conventions, workflow and more.
- [ ] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [ ] Document changes.
- [ ] Add tests for the change
- [ ] Pass `make installcheck`
- [ ] Pass `make -C src/test installcheck-cbdb-parallel`
- [ ] Feel free to request `cloudberrydb/dev` team for review and approval when your PR is ready🥳

### CODE REVIEW

File: src/include/nodes/parsenodes.h [hunk 1]
```diff
@@ -3352,6 +3407,14 @@ typedef struct CreateDirectoryTableStmt
 	char	   *tablespacename;
 } CreateDirectoryTableStmt;
```
[the remark this answers is not in the document]
Comment (wenchaozhang-123) [reply src/include/nodes/parsenodes.h:3408]: OK

File: src/test/regress/input/tag.source [hunk 2]
```diff
@@ -0,0 +1,626 @@
+-- Test tag manipulation
```
[the remark this answers is not in the document]
Comment (wenchaozhang-123) [reply src/test/regress/input/tag.source:1]: Have added single mode test case.

File: src/backend/commands/event_trigger.c [hunk 3]
```diff
@@ -1003,6 +1003,8 @@ EventTriggerSupportsObjectType(ObjectType obtype)
 			return true;
 		case OBJECT_RESQUEUE:
 		case OBJECT_RESGROUP:
+		case OBJECT_TAG:
+//		case OBJECT_TAG_DESCRIPTION:
```
[the remark this answers is not in the document]
Comment (wenchaozhang-123) [reply src/backend/commands/event_trigger.c:1007]: done.

File: src/backend/catalog/aclchk.c [hunk 4]
```diff
@@ -5647,6 +5656,33 @@ pg_opfamily_ownercheck(Oid opf_oid, Oid roleid)
...
+	if (superuser_arg(roleid))
+		return true;
+
+	tuple = SearchSysCache1(TAGOID, ObjectIdGetDatum(tag_oid));
+	if (!HeapTupleIsValid(tuple))
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("tag with OID %u does not exist",
+						tag_oid)));
+
```
[the remark this answers is not in the document]
Comment (wenchaozhang-123) [reply src/backend/catalog/aclchk.c:5678]: solved.

[the remark this answers is not in the document]
Comment (wenchaozhang-123) [reply src/backend/catalog/aclchk.c:5678]: solved.

File: src/bin/pg_dump/pg_dumpall.c [hunk 5]
```diff
@@ -47,6 +47,8 @@ static void dumpProfiles(PGconn *conn);
 static void dumpPasswordHistory(PGconn *conn);
 static void dropTablespaces(PGconn *conn);
 static void dumpTablespaces(PGconn *conn);
+static void dumpTags(PGconn *conn);
+//static void dumpTagDescriptions(PGconn *conn);
```
[the remark this answers is not in the document]
Comment (wenchaozhang-123) [reply src/bin/pg_dump/pg_dumpall.c:51]: solved.

File: src/include/nodes/parsenodes.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (wenchaozhang-123) [reply src/include/nodes/parsenodes.h:3408]: No, directory table supports tags while CATS does not.

**Review by lss602726449 — APPROVED** [review]

**Review by my-ship-it — APPROVED** [review]

**Pull request merged**